src/ops/base.rs

```diff
@@ -1,39 +1,221 @@
+    // unique within the same process, the assignment of ids is not deterministic across multiple
+    // process runs. This is just a tuple of 2 monotonically increasing counters: the first is per
+    // process, and the second is "within" that process.
+    unique_id: ProcessUniqueId,
+
     us: Option<NodeAddress>,
 }
 
+/// Specifies the level of durability that this base node should offer. Stronger guarantees imply a
+/// reduced write performance.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum BaseDurabilityLevel {
+    /// No durability at all: records aren't written to a log. This is the default durability level
+    /// if none is specified.
+    None,
+    /// Buffered writes: records are accumulated in an in-memory buffer and occasionally flushed to
+    /// the durable log, which may itself buffer in the file system. Results in large batched
+    /// writes, but offers no durability guarantees on crashes.
+    Buffered,
+    /// Synchronous writes: forces every record to be written to disk before it is emitted further
+    /// into the data-flow graph. Strong guarantees (writes are never lost), but high performance
+    /// penalty.
+    SyncImmediately,
+}
+
 impl Base {
     /// Create a base node operator.
-    pub fn new(primary_key: Vec<usize>) -> Self {
+    pub fn new(primary_key: Vec<usize>, durability: BaseDurabilityLevel) -> Self {
         Base {
+            buffered_writes: Some(Records::default()),
+            durability: durability,
+            durable_log: None,
+            durable_log_path: None,
+            global_address: None,
+            last_flushed_at: Some(Instant::now()),
             primary_key: Some(primary_key),
+            should_delete_log_on_drop: false,
+            unique_id: ProcessUniqueId::new(),
             us: None,
         }
     }
+
+    /// Whether this base node should delete its durable log on drop.  Used when durable log is not
+    /// intended to be used for future recovery, e.g., on tests.
+    pub fn delete_log_on_drop(mut self) -> Base {
+        self.should_delete_log_on_drop = true;
+        self
+    }
+
+    /// Write records to durable log.
+    fn persist_to_log(&mut self, records: &Records) {
+        match self.durability {
+            BaseDurabilityLevel::None => panic!("tried to persist non-durable base node!"),
+            BaseDurabilityLevel::Buffered |
+            BaseDurabilityLevel::SyncImmediately => {
+                self.ensure_log_writer();
+                serde_json::to_writer(&mut self.durable_log.as_mut().unwrap(), &records)
+                    .unwrap();
+                // XXX(malte): we must deconstruct the BufWriter in order to get at the contained
+                // File (on which we can invoke sync_data(), only to then reassemble it
+                // immediately. I suspect this will work best if we flush after accumulating
+                // batches of writes.
```

Comment:
It does, but we still need the original `file` to create a the new `BufWriter`.  So we can instead drop `file.sync_data()` -- since we need the `file` anyway on the next line.